Source/Dafny/Translator.cs

```diff
@@ -9891,6 +9891,13 @@ void TrStmt(Statement stmt, BoogieStmtListBuilder builder, List<Variable> locals
             builder.Add(TrAssumeCmd(stmt.Tok, etran.TrExpr(s.Expr)));
             stmtContext = StmtType.NONE;
           }
+        } else if (stmt is ExpectStmt) {
+          AddComment(builder, stmt, "expect statement");
+          ExpectStmt s = (ExpectStmt)stmt;
+          stmtContext = StmtType.ASSUME;
+          TrStmt_CheckWellformed(s.Expr, builder, locals, etran, false);
+          builder.Add(TrAssumeCmd(stmt.Tok, etran.TrExpr(s.Expr), etran.TrAttributes(stmt.Attributes, null)));
```

Comment:
This gives the translation

    CheckWellformed[[ s.Expr ]]
    assume TrExpr[[ s.Expr ]]

but it also needs to check that `s.Message` is well-formed _if the condition does not hold_. This is important, because otherwise the program could crash when evaluating .Message, when it should have halted.

So, instead do:

    CheckWellformed[[ s.Expr ]]
    if TrExpr[[ s.Expr ]] {
    } else {
      CheckWellformed[[ s.Message ]]
      assume false;
    }

Also, add some tests cases that demonstrate that `.Message` is checked for well-formedness under the assumption that the condition does _not_ hold.

Reply:
Excellent catch! Addressed in the most recent commit.